### PR TITLE
os/bluestore/BlueFS: support write time hint.

### PR DESCRIPTION
**Pull request by majianpeng**
Now rocksdb support this feature in io_posix.cc. This feature make SSD better understand the data life and based on this to do prefer allocate and do correctly GC. 
In kernel, this feature based on file or inode. For BlueFs, it use the same raw device for multi files. So i open multi rad-write for different write-hint. Then for write based on hint to choose fd.

### CODE REVIEW

**Comment by majianpeng** [comment]
@liewegas . please review.

**Comment by majianpeng** [comment]
@liewegas . ping

**Comment by majianpeng** [comment]
@liewegas . Have you time to review this? Thanks!

File: src/os/bluestore/BlueRocksEnv.cc [hunk 1]
```diff
@@ -227,6 +227,13 @@ class BlueRocksWritableFile : public rocksdb::WritableFile {
     return false;
   }
 
+  void SetWriteLifeTimeHint(rocksdb::Env::WriteLifeTimeHint hint) override {
+    if (hint == h->write_hint) {
```
Comment (ifed01) [inline src/os/bluestore/BlueRocksEnv.cc]: why do we need this check?

File: src/os/bluestore/KernelDevice.cc [hunk 2]
```diff
@@ -47,6 +45,12 @@ KernelDevice::KernelDevice(CephContext* cct, aio_callback_t cb, void *cbpriv, ai
     discard_thread(this),
     injecting_crash(0)
 {
+  fd_directs.resize(WRITE_LIFE_MAX);
```
Comment (ifed01) [inline src/os/bluestore/KernelDevice.cc]: you can use vector::resize(size, val) for to initialize here

File: src/os/bluestore/KernelDevice.cc [hunk 3]
```diff
@@ -64,23 +68,45 @@ int KernelDevice::_lock()
...
-  if (fd_buffered < 0) {
-    r = -errno;
-    derr << __func__ << " open got: " << cpp_strerror(r) << dendl;
-    goto out_direct;
+  if (i != WRITE_LIFE_MAX) {
+      derr << __func__ << " open got: " << cpp_strerror(r) << dendl;
+      goto out_fail;
   }
+
+  for (i = 0; i < WRITE_LIFE_MAX; i++) {
```
Comment (ifed01) [inline src/os/bluestore/KernelDevice.cc]: Probably start from WRITE_LIFE_NONE

File: src/os/bluestore/KernelDevice.cc [hunk 4]
```diff
@@ -64,23 +68,45 @@ int KernelDevice::_lock()
...
+    if (fcntl(fd_directs[i], F_SET_FILE_RW_HINT, &i) < 0) {
```
Comment (liewegas) [inline src/os/bluestore/KernelDevice.cc]: Will this fail on old kernels that we care about?  Should we tolerate EOPNOTSUPP or similar?

Comment (majianpeng) [reply src/os/bluestore/KernelDevice.cc]: If failed on old kernel, we disable enable_wrt and later in choose_fd, we only use fd_directs/fd_buffered[WRITE_LIFE_NOT_SET].

**Pull request merged**